text/0000-unsafe-fields.md

```diff
@@ -0,0 +1,137 @@
+- Feature Name: `unsafe_fields`
+- Start Date: 2023-07-13
+- RFC PR: [rust-lang/rfcs#0000](https://github.com/rust-lang/rfcs/pull/0000)
+- Rust Issue: [rust-lang/rust#0000](https://github.com/rust-lang/rust/issues/0000)
+
+# Summary
+
+Fields may be declared `unsafe`. Unsafe fields may only be mutated (excluding interior mutability)
```

Comment:
Why exclude interior mutability? That should be explained in the RFC.

Reply:
It's addressed in the linked RFC that explains what is considered a mutable use.

Reply:
I don't see how that RFC shows why specifically for unsafe fields we should ignore interior mutability. If I have an invariant involving a `Cell<i32>`, surely calling `set` on it should require an unsafe block?

Reply:
> If interior mutability were included,vit would require that any reference (as in immutable reference) to a !Freeze type be considered unsafe

Indeed that is what I would expect. It seems less undesirable than having `field.set(...)` violate an invariant.

Reply:
`!Freeze` isn't sufficient because `Box<Cell<T>>: !Freeze`. This would require creating any reference to unsafe fields to be considered unsafe. Only a move/copy would be safe.

Reply:
> !Freeze isn't sufficient because Box<Cell<T>>: Freeze. This would require creating any reference to unsafe fields to be considered unsafe. Only a move/copy would be safe.

Ah, good point. Not entirely sure how I feel about that.

Still at the very least the RFC should discuss interior mutability and mention the alternative of also attempting to prevent mutation of interior mutable unsafe fields.

(It already is a breaking change, as interior mutability matters for consts, statics, implicit promotion.)

Reply:
I suppose it's not that big of a deal to rip the text from the previous RFC, adapting slightly as appropriate. I can elaborate on interior mutability specifically.